zkvm/src/types.rs

```diff
@@ -59,6 +59,12 @@ pub enum Data {
 
     /// An input object (claimed UTXO).
     Output(Box<Output>),
+
+    /// A call proof.
+    CallProof(CallProof),
+
+    /// A program witness.
+    ProgramWitness(ProgramWitness),
```

Comment:
There's some redundancy here: `Data::Opaque` and `Data::Program` already support verifier's and prover's representations of the program, respectively. What if we kept only them, and had `data.to_program_witness()` method that converts `Data::Opaque->ProgramWitness::Opaque` and `Data::Program->ProgramWitness::Program`, and fails otherwise. Then it still would avoid premature parsing, and there would be no redundancy. 

We already do the similar dual-downcast of Opaque/Nonopaque types with `Data::to_predicate` and `Data::to_commitment` - check them out.